File: go/jit/internal/tuple.go
/*
 * onejit - JIT compiler in Go
 *
 * Copyright (C) 2018-2020 Massimiliano Ghilardi
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * tuple.go
 *
 *  Created on Dec 20, 2020
 *      Author Massimiliano Ghilardi
 */

package internal

type TupleExpr struct {
	op   Op
	list []Expr
}

func Tuple(op Op, expr ...Expr) *TupleExpr {
	return tuple(op, dup(expr))
}

func Ret(expr ...Expr) *TupleExpr {
	return tuple(RET, dup(expr))
}

func AssignCall(to []Expr, call *CallExpr) *TupleExpr {
	return tuple(ASSIGN, append(dup(to), call))
}

// implement Expr interface
func (e *TupleExpr) Class() Class {
	return TUPLE
}

func (e *TupleExpr) Op() Op {
	return e.op
}

func (e *TupleExpr) Kind() Kind {
	return Void
}

func (e *TupleExpr) expr() {}

func (e *TupleExpr) RegId() RegId {
	return NoRegId
}

func (e *TupleExpr) Size() Size {
	return Void.Size()
}

func (e *TupleExpr) Children() int {
	return len(e.list)
}

func (e *TupleExpr) Child(i int) Node {
	return e.list[i]
}

func (e *TupleExpr) IsConst() bool {
	return false
}

func (e *TupleExpr) IsPure() bool {
	// tuples are either assignments or returns => they have side effects
	return false
}

// ========================= helpers ===========================================

func dup(expr []Expr) []Expr {
	ret := make([]Expr, len(expr))
	copy(ret, expr)
	return ret
}

func tuple(op Op, expr []Expr) *TupleExpr {
	switch op {
	case ASSIGN:
		n := len(expr) - 1
		if n < 0 {
			Errorf("missing CallExpr in tuple %v", op)
		}
		call, ok := expr[n].(*CallExpr)
		if !ok {
			Errorf("bad last expression in tuple %v: have %T, want CallExpr", op, expr[n])
		}
		sig := call.Signature()
		if n != sig.NumOut() {
			Errorf("bad number of destinations in tuple %v: have %d, want %d", op, n, sig.NumOut())
		}
		for i := 0; i < n; i++ {
			kind := expr[i].Kind()
			karg := sig.Out(i)
			if karg != kind {
				Errorf("bad destination %v in tuple %v: have %d, want %d", i+1, op, karg, kind)
			}
		}
		break
	case RET:
		break
	default:
		Errorf("invalid tuple operation: %v", op)
	}
	return &TupleExpr{
		op:   op,
		list: expr,
	}
}

File: onejit/tuple.cpp
/*
 * onejit - JIT compiler in C++
 *
 * Copyright (C) 2018-2021 Massimiliano Ghilardi
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * tuple.cpp
 *
 *  Created on Jan 28, 2021
 *      Author Massimiliano Ghilardi
 */

#include <onejit/code.hpp>
#include <onejit/func.hpp>
#include <onejit/functype.hpp>
#include <onejit/label.hpp>
#include <onejit/mem.hpp>
#include <onejit/tuple.hpp>
#include <onestl/buffer.hpp>

namespace onejit {

// ============================  Tuple  ====================================

Node Tuple::create(Func &func, Kind kind, OpN op, Nodes nodes) noexcept {
  const size_t n = nodes.size();
  Code *holder = func.code();

  while (holder && n == uint32_t(n)) {
    const NodeHeader header{TUPLE, kind, op};
    CodeItem offset = holder->length();

    if (holder->add(header) && holder->add_uint32(n) && holder->add(nodes, offset)) {
      return Node{header, offset, holder};
    }
    holder->truncate(offset);
    break;
  }
  return Node{};
}

const Fmt &operator<<(const Fmt &out, const Tuple &expr) {
  if (expr.type() == MEM) {
    // may happen, Mem is a subclass of Tuple
    return out << expr.is<Mem>();
  }

  OpN op = expr.op();
  out << '(' << op;
  const bool is_call = op == CALL;
  // if op == CALL, skip child(0) i.e. FuncType
  for (size_t i = size_t(is_call), n = expr.children(); i < n; i++) {
    Node child = expr.child(i);
    if (child) {
      out << ' ' << child;
    }
  }
  return out << ')';
}

} // namespace onejit
